### DISCUSSION TITLE
Linking c/c++ executable to shared libs

### DISCUSSION THREAD

**Discussion by ericgtkb** [mesonbuild/meson-python#638] [Q&A] [answered]
I have a project where I am building python extensions as well as some other c/c++ executables.
The build script is roughly as follows.
```meson
project('myproj', 'c', version: '1.0')

py = import('python').find_installation(pure: false)

# Shared lib
mylib = library('mylib', 'my_source.c', install: true)

# Python extension module
py.extension_module('_pyext', 'pyext.c', link_with: mylib, install: true, subdir: 'pyext')
py.install_sources('__init__.py', subdir: 'pyext')

# C executable
executable('myexe', 'main.c', link_with: mylib, install: true)
```
When running pip install mesonpy maps `bindir` to python's `script` directory (`venv/bin` in my case), and `libdir` to `.myproj.mesonpy.libs` (`venv/lib/python.X.Y/site-packages/.myproj.mesonpy.libs` in my case). For the python extension this is all fine and it is able to correctly link to `libmylib.so` in `venv/lib/python.X.Y/site-packages/.myproj.mesonpy.libs`, but the C executable `myexe` (in `venv/bin`) is not correctly linked.

Is there anyway, to correct the rpath for the C executable so it links to `libmylib.so` correctly in `venv/lib/python.X.Y/site-packages/.myproj.mesonpy.libs`? I've tried using the `install_rpath` option with `libdir`, but it still not able to set the correct rpath.

**Comment by rgommers** [comment, marked as answer]
Hi @ericgtkb, thanks for the clear example. 

My first thought is that it's probably better to do one of two things:
1. make `mylib` a static library and link it into the executable
2. install both the shared library and the executable inside `site-packages/myproj` in the same subdirectory. Then add a `[project.scripts]` entrypoint in `pyproject.toml` to call the executable

(1) is an easier and smaller change than (2) if that's acceptable. If you do need the shared library installed, then you need (2). Both are portable also to Windows, while setting an RPATH is not. I'm also not even sure if the relative path from {bindir} to site-packages is constant for all possible install schemes.

**Reply by rgommers** [reply]
Note that to install a shared library in a different location inside `site-packages/myproj`, you need to add `install_dir: py3.get_install_dir() / 'myproj'`,

**Reply by ericgtkb** [reply]
Thanks for the suggestions! Making `mylib` static is probably the best option for my project.

**Reply by Yashdate10** [reply]
Will you please suggest me some resources to learn DSA in C++?
